controllers/array_action/array_mediator_svc.py

```diff
@@ -1545,34 +1559,32 @@ def _is_vdisk_has_fcmaps(self, vdisk_uid):
         cli_volume = self._get_cli_volume_by_wwn(vdisk_uid, not_exist_err=False)
         return cli_volume and cli_volume.FC_id
 
-    def _mkhost(self, host_name, connectivity_type, ports):
-        cli_kwargs = build_create_host_kwargs(host_name, connectivity_type, ports)
+    def _is_port_invalid(self, message_from_storage):
+        return FC_PORT_IS_NOT_VALID in message_from_storage or \
+            ISCSI_PORT_IS_NOT_VALID in message_from_storage or \
+            NVME_PORT_IS_ALREADY_ASSIGNED in message_from_storage
+
+    def _mkhost(self, host_name, connectivity_type, port):
+        cli_kwargs = build_create_host_kwargs(host_name, connectivity_type, port)
         try:
             self.client.svctask.mkhost(**cli_kwargs)
+            return 200
         except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
-            if is_warning_message(ex.my_message):
-                logger.warning("exception encountered during host {} creation : {}".format(host_name, ex.my_message))
             if OBJ_ALREADY_EXIST in ex.my_message:
                 raise array_errors.HostAlreadyExists(host_name, self.endpoint)
+            if self._is_port_invalid(ex.my_message):
+                return 400
+            if is_warning_message(ex.my_message):
+                logger.warning("exception encountered during host {} creation : {}".format(host_name, ex.my_message))
             raise ex
 
-    def _get_connectivity_type_by_initiators(self, initiators):
-        if initiators.nvme_nqns:
-            return array_settings.NVME_OVER_FC_CONNECTIVITY_TYPE
-        if initiators.fc_wwns:
-            return array_settings.FC_CONNECTIVITY_TYPE
-        if initiators.iscsi_iqns:
-            return array_settings.ISCSI_CONNECTIVITY_TYPE
-        return None
-
     def create_host(self, host_name, initiators, connectivity_type):
-        if not connectivity_type:
-            connectivity_type = self._get_connectivity_type_by_initiators(initiators)
-        ports = initiators.get_by_connectivity_type(connectivity_type)
-        if ports:
-            self._mkhost(host_name, connectivity_type, ports)
-            return
-        raise array_errors.NoPortFoundByConnectivityType(initiators, connectivity_type)
+        ports = get_connectivity_type_ports(initiators, connectivity_type)
```

Comment:
so `connectivity_type` is a mandatory field now?

Reply:
yes I moved the `connectivity_type` logic to the servicer because I needed the `connectivity_type` before calling the mediators and because it's not relate to a specific mediator (even thou we implemented `create_host` only in one mediator)